Framework/Indexing/src/IndexInfo.cpp

```diff
@@ -178,25 +148,37 @@ SpectrumIndexSet IndexInfo::makeIndexSet(
  * subset of indices on this partition. */
 SpectrumIndexSet IndexInfo::makeIndexSet(
     const std::vector<GlobalSpectrumIndex> &globalIndices) const {
-  makeSpectrumNumberTranslator();
   return m_spectrumNumberTranslator->makeIndexSet(globalIndices);
 }
 
-void IndexInfo::makeSpectrumNumberTranslator() const {
-  // To support legacy code that creates workspaces with duplicate spectrum
-  // numbers we are currently creating the SpectrumNumberTranslator only when
-  // needed.
-  // Note that the lazy initialization implies that no sharing of
-  // SpectrumNumberTranslator will be set up for workspaces created before the
-  // first access.
-  // TODO We are not setting monitors currently. This is ok as long as we have
-  // exactly one partition.
-  PartitionIndex partition = 0;
-  RoundRobinPartitioner partitioner(
-      1, partition, Partitioner::MonitorStrategy::CloneOnEachPartition,
-      std::vector<GlobalSpectrumIndex>{});
+bool IndexInfo::isOnThisPartition(GlobalSpectrumIndex globalIndex) const {
+  // A map from global index to partition might be faster, consider adding this
+  // if it is used a lot and has performance issues.
+  const auto helperSet = makeIndexSet(globalIndex, globalIndex);
+  return helperSet.size() == 1;
+}
+
+void IndexInfo::makeSpectrumNumberTranslator(
+    std::vector<SpectrumNumber> &&spectrumNumbers) const {
+  PartitionIndex partition;
+  int numberOfPartitions;
+  if (m_storageMode == StorageMode::Distributed) {
+    partition = PartitionIndex(m_communicator.rank);
+    numberOfPartitions = m_communicator.size;
+  } else if (m_storageMode == StorageMode::Cloned) {
```

Comment:
I am confused here. Is it not possible to have a cloned storage mode with multiple partitions? I thought the data would be duplicated on each partition.